### PR TITLE
Cleaning up and covering the best practices with tests

### PR DESCRIPTION
**Pull request by andreasohlund**
The checks for send only was removed:

https://github.com/Particular/NServiceBus/commit/d858cadd319c4f27c019f88be9dfc355143ae5fa#diff-15cf998feca08db309a1930e0f7bbd3aL213

since users can't call subscribe/unsubscribe for send only endpoints since the `ISendOnlyBus` doesn't expose those calls

### Documentation

https://github.com/Particular/docs.particular.net/pull/617

### CODE REVIEW

File: src/NServiceBus.Core/v6-obsoletes.cs [hunk 1]
```diff
@@ -110,6 +110,12 @@ public abstract class DeliveryOptions
            RemoveInVersion = "7.0",
            TreatAsErrorFromVersion = "6.0")]
         public string ReplyToAddress { get { throw new NotImplementedException(); } set { throw new NotImplementedException(); } }
+
+        [ObsoleteEx(
+           Message = "Turn best practices check off using configuration.DisableFeature<BestPracticesEnforcement>()",
```
Comment (johnsimons) [inline src/NServiceBus.Core/v6-obsoletes.cs]: `configuration.DisableFeature<BestPracticesEnforcement>()` not very discoverable IMO, should we add an extension to busconfiguration instead ?

**Comment by johnsimons** [comment]
If I correctly understand this change, this means that we now have an endpoint instance toggle switch for `EnforceMessagingBestPractices`, so it is not a per message decision, was this the intended result @andreasohlund ?

Shouldn't it be a per message decision, one scenario that comes to mind is interop, or migrating from old to new ?
Also, I thought we needed this for the gateway `SendToSites` ?

File: src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            EnableByDefault();
+        }
+        /// <summary>
+        /// Initialized the feature
+        /// </summary>
+        /// <param name="context">The feature context</param>
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            context.Container.ConfigureComponent<Validations>(DependencyLifecycle.InstancePerCall);
```
Comment (johnsimons) [inline src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs]: It looks like `Validations` is a stateless class, should it be registered as a singleton

File: src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+namespace NServiceBus.Features
+{
+    using NServiceBus.MessagingBestPractices;
+    using NServiceBus.Pipeline;
+
+    /// <summary>
+    /// Makes sure that messaging best practices are followed
+    /// </summary>
+    public class BestPracticesEnforcement : Feature
```
Comment (johnsimons) [inline src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs]: You [pick on my class naming](https://github.com/Particular/NServiceBus/pull/2689#discussion_r29410570), and now you create a pluralized class :wink:

**Comment by johnsimons** [comment]
@andreasohlund isn't AT a bit too heavy to test arguments validation ?

**Comment by andreasohlund** [comment]
>isn't AT a bit too heavy to test arguments validation ?

I was on the fence, but there is no good way to test the check for subscribe/unsubscribe. I still felt it was worth it. Now that I'm about to add a override per message I'd say it's worth the tax to keep them

File: src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs]: Will fix

File: src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs (hunk 3 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/MessagingBestPractices/BestPracticesEnforcement.cs]: Lol, you're correct. I'll rename

**Comment by andreasohlund** [comment]
After talking to @johnsimons we decided to add the ability to turn this off per message as well

**Pull request merged**